config/module.go

```diff
@@ -233,76 +241,170 @@ func (m Module) EvaluateExePath(packagesDir string) (string, error) {
 // could break or uncoordinate package sync.
 const FirstRunSuccessSuffix = ".first_run_succeeded"
 
-// EvaluateFirstRunPath returns absolute FirstRunPath from one of two sources (in order of precedence):
-// 1. if there is a meta.json in the exe dir, use that, except in local non-tarball case.
-// 2. if this is a local tarball and there's a meta.json next to the tarball, use that.
-// Note: the working directory must be the unpacked tarball directory or local exec directory.
-//
-// On success (i.e. if the returned error is nil), this function also returns a function that creates
-// a marker file indicating that the setup phase has run successfully.
-func (m Module) EvaluateFirstRunPath(packagesDir string, logger logging.Logger) (
-	string,
-	func() error,
-	error,
-) {
-	noop := func() error { return nil }
-	unpackedModDir, err := m.exeDir(packagesDir)
+// FirstRun executes a module-specific setup script.
+func (m *Module) FirstRun(
+	ctx context.Context,
+	localPackagesDir,
+	dataDir string,
+	env map[string]string,
+	logger logging.Logger,
+) error {
+	logger = logger.Sublogger("first_run").WithFields("module", m.Name)
+
+	unpackedModDir, err := m.exeDir(localPackagesDir)
 	if err != nil {
-		return "", noop, err
+		return err
 	}
 
+	// check if FirstRun already ran successfully for this module version by
+	// checking if a success marker file exists on disk.
 	firstRunSuccessPath := unpackedModDir + FirstRunSuccessSuffix
 	if _, err := os.Stat(firstRunSuccessPath); !errors.Is(err, os.ErrNotExist) {
 		logger.Info("first run already ran")
-		return "", noop, errors.New("first run already ran")
+		return nil
+	}
+
+	// Load the module's meta.json. If it doesn't exist or fails to load, debug log and exit quietly.
+	meta, err := m.getJSONManifest(unpackedModDir)
+	if err != nil {
+		logger.Debugw("failed to load meta.json, skipping setup phase", "error", err)
```

Comment:
```suggestion
		logger.Debugw("failed to load meta.json, skipping first run phase", "error", err)
```
Unsure what messaging should be to users, but I'm assuming `setup` was old name for `first run`.